### ISSUE TITLE
Update Google API PHP client services library to version with Idea Hub and AdSense V2 API

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#3517]
Now that the [auto-generated PHP client code for the Idea Hub API](https://github.com/googleapis/google-api-php-client-services/blob/master/src/Google/Service/Ideahub.php) is available, we need to upgrade the Google API services library to include it.

The main reason for the update is the above Idea Hub API - however, related to that, the new PHP client services library also includes the new AdSense API client code, which switches from the v1.4 API to the v2 API - **with the v1.4 API getting deprecated in October this year**. The changes are minimal, from an API perspective, only the API root differs (see https://developers.google.com/adsense/management/migrating). However, the client code appears to be generated on a slightly different foundation, so some of Site Kit's usage of it needs to be adjusted. The changes are purely naming- and format-related though, no logic changes.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The `google/apiclient-services` PHP library in `composer.json` should be updated to `^0.180` (latest version, which includes Idea Hub), and it should be updated accordingly (in `composer.lock` as well).
* The list of Google services to include the PHP third-party code for in the PHP scoper config `scoper.inc.php` should be modified to include `Ideahub`.
* The Google service classes for AdSense from the library now use `Adsense` instead of `AdSense` in their name, so this needs to be adjusted in `scoper.inc.php` as well.
* All usage of the AdSense PHP service classes also needs to be reviewed to follow the new names that the library code uses now.

## Implementation Brief

* Update `composer.json` as outlined in the ACs, then run `composer update google/apiclient-services` to actually update it.
* Modify `scoper.inc.php` to add `Ideahub` and replace `AdSense` with `Adsense`.
* Update code in `AdSense` module class to work with the new PHP library client code:
    * Update all library class name references from `Google_Service_AdSense` to `Google_Service_Adsense`.
    * Some resources have been renamed to more precisely follow their hierarchy, e.g. `accounts_urlchannels` is now `accounts_adclients_urlchannels` and `accounts_adunits` is now `accounts_adclients_adunits`. This applies both for the resource property name as well as for the resource class name.
    * Any parent IDs that need to be specified now must no longer be passed as individiual parameters, but as a single `$parent` parameter, which works similarly to e.g. the Google Analytics 4 Admin API already does (see `Analytics_4` class), specifying an account ID as `accounts/{accountID}` and an ad client ID as `accounts/{accountID}/adclients/{adclientID}`.
    * The only endpoint that includes additional changes in the request format is the one for generating a report; follow the [documentation](https://developers.google.com/adsense/management/reference/rest/v2/accounts.reports/generate) and migrate the existing arguments to be passed in that format.
    * The returned objects from the API also slightly differ, for example they now no longer include the IDs as properties, but use a `name` property; so we'll now need to also do a similar ID extraction in the `AdSense` module class that is already present in the `Analytics_4` class, adding the IDs in underscore-prefixed extra properties (e.g. `_accountID` on an account object).
* Update all references to IDs and other values in objects returned from the AdSense API (most commonly change `id` to `_id` and `name` to `displayName`) to refer to the now manually added, underscore-prefixed ID properties.
* Review all AdSense Storybook components to ensure they still receive the correct data and continue to work.
* Complete and merge #3548.

### Test Coverage

* N/A

### Visual Regression Changes

* N/A

## QA Brief

* Test the AdSense module thoroughly, basically to ensure it still works as expected (since this issue switched all the codebase around AdSense to its new v2 API). For example covering:
    * the AdSense setup flow, with different scenarios in the account
        * To mock certain account and site states, you will need to use version 1.6.0 of the tester plugin which has been updated to work with AdSense API v2 (i.e. this PR and Site Kit itself once this PR is merged). That version will be published shortly (you should see an update prompt in your WP site with it by Wednesday, June 16).
        * Mocking data should only be used to test all different states. Some basic testing for the AdSense setup should also be done using a real AdSense account (i.e. without the tester plugin).
    * the AdSense widgets in the dashboard (test with real AdSense data, i.e. without the tester plugin)
    * the AdSense module screen

## Changelog entry

* Update API client library and AdSense integration to use the new version 2 of their API.

**Comment by felixarntz** [comment]
@aaemnnosttv The scope here increased unfortunately. This was supposed to be a very simple issue, but it also comes with the updated version of the AdSense API code, which now uses v2 of the AdSense API. This is not rocket science, but requires a bunch of code changes (see IB). We need to update to the new version of the API soon anyway though, and since it is now blocking Idea Hub, this needs to be prioritized. (The Idea Hub work in here is so minimal that it doesn't make sense to split this into two issues.)

@eclarke1 @fhollis I've already started to work on a PR for this issue, so I'll continue to tackle this one and get it to code review - just pointing it out for awareness that the execution-related bit of the estimate will be mostly on me.

**Comment by felixarntz** [comment]
@eclarke1 @fhollis @aaemnnosttv I've made quite some progress with the PR - I almost wanna say it looks complete, but probably a few things will still come up from testing. I've moved this to Sprint 51, since it blocks various other Idea Hub issues, and I'm going to prioritize completing it. Let's just make sure to not merge it until the upcoming release has been branched off, since this one should be around for a while before a release, to get as much natural testing as possible.

**Comment by aaemnnosttv** [comment]
IB ✅

**Comment by felixarntz** [comment]
Quick update here @aaemnnosttv: The PR is in principle ready for review, but there are a few details left to clarify internally with the AdSense API team, specifically whether the alert type and error reason values that we've been relying on still exist like that in the new API.

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@felixarntz @aaemnnosttv With a site that has live Adsense data (no tester plugin activated) For 7, 14 and 90 days, there's no chart appearing.  For 28 days, the chart does appear and you can see that there's data for these periods, be it small amounts.  

![image](https://user-images.githubusercontent.com/73545194/123287658-cc5b7d00-d506-11eb-935b-f76c4abfaf4b.png)

https://user-images.githubusercontent.com/73545194/123287877-fd3bb200-d506-11eb-8082-abb78946b972.mp4

**Comment by felixarntz** [comment]
@wpdarren Thanks! Do you see any (request) errors in the JavaScript console in devtools when changing the date ranges to those values where it doesn't return anything?

**Comment by wpdarren** [comment]
@felixarntz interestingly, when I have gone into the same test site, no data appears for any period now, and a number of errors appears as per the screenshot.  I am unsure if this is due to the site I am using.  Please let me know if you'd like to see all the errors, or, try test this without another website with Adsense data. 

![image](https://user-images.githubusercontent.com/73545194/123297478-82c36000-d50f-11eb-970d-77568b3b0fbb.png)

**Comment by felixarntz** [comment]
@wpdarren Ahhh thanks. I see, in [the metric docs](https://developers.google.com/adsense/management/reference/rest/v2/Metric) it looks like some of the supported metric values have changed in the AdSense API v2. I'll create a follow-up PR for you to test.

**Comment by felixarntz** [comment]
@wpdarren Can you re-test this again with PR #3630 please? I've updated the metrics Site Kit uses there accordingly. Please test without the AdSense mocks from the tester plugin, since I'll need to update it there as well.

**Comment by wpdarren** [comment]
@felixarntz this is looking better, we have data for 7, 14, 28 and 90 days now.  

I have a few observations:

1.  We have minus data in the charts, i.e. for previous period, in earnings. I am assuming this could happen with Adsense, and could be linked to [3564](https://github.com/google/site-kit-wp/issues/3564). Wasn't 100% sure so reporting it. 
2. The dates seem to be duplicated on the x-axis.  The period is 7 days.  All other periods display fine. 

![image](https://user-images.githubusercontent.com/73545194/123449876-fcbd1d00-d5d3-11eb-8139-487bbb44e0e7.png)

I am continuing on with the testing, but thought I would mention these observations now.

**Comment by aaemnnosttv** [comment]
> We have minus data in the charts, i.e. for previous period, in earnings. I am assuming this could happen with Adsense. But it could also be linked to [3564](https://github.com/google/site-kit-wp/issues/3564). Wasn't 100% sure so reporting it.

@wpdarren This would be due to #3564 indeed 👍 

> The dates are duplicated on the x-axis.  The period is 7 days.  All other periods display fine. 

Is this not the case with the current version? I've seen that before but don't recall if we did something do change it or not; it shouldn't really change related to this issue though.

**Comment by felixarntz** [comment]
@wpdarren The negative graph is the same problem as already defined in #3564, so that should cover that bit.

Regarding the duplicate dates, this looks like it's actually an unrelated problem not caused by this issue. It's also happening on other graphs when using a small date range like last 7 days. Can you open a dedicated issue to fix this throughout all graphs though please? Thanks!

Assigning back to you for confirmation, but we can probably move this to Approval if there is nothing else.

**Comment by eclarke1** [comment]
We have added [#3564](https://github.com/google/site-kit-wp/issues/3564) to sprint 52 (the current sprint) as it was in CR anyway.

**Comment by wpdarren** [comment]
### QA Update: Pass ✅ 

- Tested the AdSense module to ensure that it works as expected, including the AdSense setup flows.
- Tested the AdSense module with real AdSense data.
- Tested the AdSense widgets in the dashboard with real AdSense data.
- While no UI/UX changes, checked the module on smaller screen sizes.

**Comment by aaemnnosttv** [comment]
> I'll create a ticket for this, but maybe this should be fixed for this release?

@wpdarren if we knew what it was from maybe we could, but it's more of a minor inconsistency than a serious bug so I wouldn't say it's blocking at all. Thanks for opening that issue 👍

**Comment by aaemnnosttv** [comment]
@felixarntz it looks like there are changes in the v2 format of the [`ReportResult`](https://developers.google.com/adsense/management/reference/rest/v2/ReportResult) which cause the module to be in a zero-data state when it isn't.

The main difference seems to be that the old format returned an array of values `string[]` and the new format returns a [`Row`](https://developers.google.com/adsense/management/reference/rest/v2/ReportResult#row) or `Row[]` which is an object `{ cells: Cells[] }`.

Not sure how this was missed but obviously all tests use the old format (and tester plugin too?) so this may only be seen using real data.

Going to reopen as a high-priority item to fix for the release as AdSense is not working properly as-is.

**Comment by wpdarren** [comment]
@aaemnnosttv @felixarntz two observations I wanted to mention: 1) when I tested this I used the alpha version of the tester plugin and the issue I reported during smoke testing did not occur, I was able to connect to AdSense. 2) I also used a site that had live data without the tester plugin, and also did not receive any error messages. Would be interested to know if possible how this error could now arise, for future reference and learning what I could have done differently.

**Comment by wpdarren** [comment]
@aaemnnosttv @felixarntz I have gone through QA tasks on the PR and data loads as expected.  Please assign this to me when it's in `main` and I'll give it another run through.

**Comment by felixarntz** [comment]
@wpdarren Back to you for QA now. Please make sure to test with real AdSense data, and particularly for the following feature flag configuration, which matches the current default experience:
* `widgets.dashboard` and `widgets.pageDashboard` enabled
* `widgets.moduleScreens` disabled

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@felixarntz @aaemnnosttv I have completed testing on a site with live AdSense data but thought I would also check through the set up flow with the tester plugin in the different account status. When set to approved in the tester plugin. The message appears `The caller does not have permission` on the `Earnings` widget within the Dashboard.  On the module page it just says `AdSense data is not yet available, please check back later` Usually displayed demo data is displayed straightaway.  Is this an issue on the tester plugin side? Don't want to assume.. 

**Verified:**

All of these tests were completed on a site with live data. 

`widgets.dashboard` and `widgets.pageDashboard` enabled
`widgets.moduleScreens` disabled

- On the dashboard, made sure that Earnings data displayed for 7, 14, 28 and 90 days. 
- On the AdSense module page, made sure that the Performance graphs loaded with data for 7, 14, 28 and 90 days. 
- Also checked to see no issues when Analytics is not linked with AdSense

`widgets.dashboard` and `widgets.pageDashboard` disabled
`widgets.moduleScreens` enabled

- On the dashboard, made sure that Earnings data displayed for 7, 14, 28 and 90 days. 
- On the AdSense module page, made sure that the Performance graphs loaded with data for 7, 14, 28 and 90 days. 
- Also checked to see no issues when Analytics is not linked with AdSense

All of these tests were completed on a site with tester plugin AdSense data enabled. 

- Set up AdSense on a new site, with the different account status, i.e. approved, pending etc.

**Comment by felixarntz** [comment]
@wpdarren The incorrect behavior with the tester plugin was expected at that point, I've just updated that plugin again and shared the version with you on Slack. Please re-test with that one.

**Comment by aaemnnosttv** [comment]
> Usually, demo data is displayed straightaway. Is this an issue on the tester plugin side?

@wpdarren As @felixarntz mentioned, the tester plugin had not been updated to account for the updated report format so this would be expected. It sounds like he's updated it now (although not published) so it should be ready for a (hopefully final) pass 😄

**Comment by wpdarren** [comment]
QA Update: Pass ✅ 

**Verified:**

All of these tests were completed on a site with live data. 

`widgets.dashboard` and `widgets.pageDashboard` enabled
`widgets.moduleScreens` disabled

- On the dashboard, made sure that Earnings data displayed for 7, 14, 28 and 90 days. 
- On the AdSense module page, made sure that the Performance graphs loaded with data for 7, 14, 28 and 90 days. 
- Also checked to see no issues when Analytics is not linked with AdSense

![image](https://user-images.githubusercontent.com/73545194/124484023-444d6100-ddab-11eb-9e23-18f2d3a0bf28.png)


`widgets.dashboard` and `widgets.pageDashboard` disabled
`widgets.moduleScreens` enabled

- On the dashboard, made sure that Earnings data displayed for 7, 14, 28 and 90 days. 
- On the AdSense module page, made sure that the Performance graphs loaded with data for 7, 14, 28 and 90 days. 
- Also checked to see no issues when Analytics is not linked with AdSense

![image](https://user-images.githubusercontent.com/73545194/124484081-562f0400-ddab-11eb-835b-6c61a8538782.png)

All of these tests were completed on a site with tester plugin AdSense data enabled. 

- Set up AdSense on a new site, with the different account status, i.e. approved, pending etc.
